fusequery/query/src/pipelines/transforms/transform_join.rs

```diff
@@ -0,0 +1,126 @@
+// Copyright 2020-2021 The Datafuse Authors.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use std::any::Any;
+use std::sync::Arc;
+
+use common_datablocks::DataBlock;
+use common_datavalues::DataSchemaRef;
+use common_datavalues::DataValue;
+use common_exception::Result;
+use common_streams::DataBlockStream;
+use common_streams::SendableDataBlockStream;
+use futures::StreamExt;
+
+use crate::pipelines::processors::IProcessor;
+
+pub struct NestedLoopJoinTransform {
+    schema: DataSchemaRef,
+    left: Arc<dyn IProcessor>,
+    right: Arc<dyn IProcessor>
+}
+
+impl NestedLoopJoinTransform {
+    pub fn try_create(
+        schema: DataSchemaRef,
+        left: Arc<dyn IProcessor>,
+        right: Arc<dyn IProcessor>
+    ) -> Result<Self> {
+        Ok(NestedLoopJoinTransform {
+            schema: schema.clone(),
+            left: left.clone(),
+            right: right.clone()
+        })
+    }
+
+    async fn read_from_sink(sink: Arc<dyn IProcessor>) -> Result<Vec<DataBlock>> {
+        let mut buffer = vec![];
+        while let Some(block) = sink.execute().await?.next().await {
+            buffer.push(block?);
+        }
+        Ok(buffer)
+    }
+}
+
+#[async_trait::async_trait]
+impl IProcessor for NestedLoopJoinTransform {
+    fn name(&self) -> &str {
+        "NestedLoopJoinTransform"
+    }
+
+    fn connect_to(&mut self, input: Arc<dyn IProcessor>) -> Result<()> {
+        Ok(())
+    }
+
+    fn inputs(&self) -> Vec<Arc<dyn IProcessor>> {
+        vec![self.left.clone(), self.right.clone()]
+    }
+
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    async fn execute(&self) -> Result<SendableDataBlockStream> {
+        let read_left_task = tokio::task::spawn(Self::read_from_sink(self.left.clone()));
```

Comment:
tokio:task::spawn will push the task to the global scheduler `thread`.
In order to limit the resources of one session, we should use `context::execute_task`:
https://github.com/datafuselabs/datafuse/blob/master/fusequery/query/src/pipelines/processors/processor_merge.rs#L64

It will bind all the threads to the context tokiio runtime.

Reply:
Thanks for reminding me about this.

I was intending to implement a really simple version at first to help me understand the whole procedure, but so far I still haven't find a proper way to build join transform into `Pipeline`.

Reply:
planner -> transform:
https://github.com/datafuselabs/datafuse/blob/master/fusequery/query/src/pipelines/processors/pipeline_builder.rs

For a sql, steps to a pipeline:
1. parser to planner
2. optimizer
3. planner -> transform by pipeline builder
4. executor the pipeline

Reply:
@BohuTANG Yes, but it seems `fusequery` currently has strong consumption that a `PlanNode` only has one `input`, which doesn't fit **join** well.

I'm finding a way to handle this properly.

Reply:
We can extend it to many if need, processor now has many inputs.